can/interfaces/pcan/pcan.py

```diff
@@ -210,50 +234,94 @@ def _recv_internal(self, timeout):
+        is_fd = (theMsg.MSGTYPE & PCAN_MESSAGE_FD.value) == PCAN_MESSAGE_FD.value
+        bitrate_switch = (theMsg.MSGTYPE & PCAN_MESSAGE_BRS.value) == PCAN_MESSAGE_BRS.value
+        error_state_indicator = (theMsg.MSGTYPE & PCAN_MESSAGE_ESI.value) == PCAN_MESSAGE_ESI.value
+        is_error_frame = (theMsg.MSGTYPE & PCAN_MESSAGE_ERRFRAME.value) == PCAN_MESSAGE_ERRFRAME.value
+
+
+        if self.fd:
+            dlc = pcan_dlc_list[theMsg.DLC]
+            timestamp = boottimeEpoch + (itsTimeStamp.value / (1000.0 * 1000.0))
+        else:
+            dlc = theMsg.LEN
+            timestamp = boottimeEpoch + ((itsTimeStamp.micros + 1000 * itsTimeStamp.millis + 0x100000000 * 1000 * itsTimeStamp.millis_overflow) / (1000.0 * 1000.0))
 
-        dlc = theMsg.LEN
-        timestamp = boottimeEpoch + ((itsTimeStamp.micros + 1000 * itsTimeStamp.millis + 0x100000000 * 1000 * itsTimeStamp.millis_overflow) / (1000.0 * 1000.0))
 
         rx_msg = Message(timestamp=timestamp,
                          arbitration_id=theMsg.ID,
-                         is_extended_id=bIsExt,
-                         is_remote_frame=bIsRTR,
+                         is_extended_id=is_extended_id,
+                         is_remote_frame=is_remote_frame,
+                         is_error_frame=is_error_frame,
                          dlc=dlc,
-                         data=theMsg.DATA[:dlc])
+                         data=theMsg.DATA[:dlc],
+                         is_fd=is_fd,
+                         bitrate_switch=bitrate_switch,
+                         error_state_indicator=error_state_indicator)
 
         return rx_msg, False
 
     def send(self, msg, timeout=None):
-        if msg.is_extended_id:
-            msgType = PCAN_MESSAGE_EXTENDED
-        else:
-            msgType = PCAN_MESSAGE_STANDARD
-
-        # create a TPCANMsg message structure
-        if platform.system() == 'Darwin':
-            CANMsg = TPCANMsgMac()
-        else:
-            CANMsg = TPCANMsg()
-
-        # configure the message. ID, Length of data, message type and data
-        CANMsg.ID = msg.arbitration_id
-        CANMsg.LEN = msg.dlc
-        CANMsg.MSGTYPE = msgType
-
-        # if a remote frame will be sent, data bytes are not important.
+        msgType = PCAN_MESSAGE_EXTENDED.value if msg.is_extended_id else PCAN_MESSAGE_STANDARD.value
         if msg.is_remote_frame:
-            CANMsg.MSGTYPE = msgType.value | PCAN_MESSAGE_RTR.value
-        else:
+            msgType |= PCAN_MESSAGE_RTR.value
+        if msg.is_error_frame:
+            msgType |= PCAN_MESSAGE_ERRFRAME.value
+        if msg.is_fd:
+            msgType |= PCAN_MESSAGE_FD.value
+        if msg.bitrate_switch:
+            msgType |= PCAN_MESSAGE_BRS.value
+        if msg.error_state_indicator:
+            msgType |= PCAN_MESSAGE_ESI.value
+
+        if self.fd:
+            # create a TPCANMsg message structure
+            if platform.system() == 'Darwin':
+                CANMsg = TPCANMsgFDMac()
```

Comment:
We could also use this change to do some small cleanup here: The usual convention for variables names is lower_case_with_underscores.

Reply:
I thought about doing some cleanup in a separate PR.
Would you rather i do some cleanup in this PR?

Reply:
Separating them is of course cleaner! I won't stop you from doing so. ;)